rse-py-package.Rmd

```diff
@@ -0,0 +1,1477 @@
+
+Also present in `dist` is a file `zipfpy-0.1-py3-none-any.whl`, the "wheel" file;
+running `file dist/zipfpy-0.1-py3-none-any.whl` will show it to be a zip file.  We
+can list its contents and see the same source files and some metadata:
+
+```shell
+$ unzip -l dist/zipfpy-0.1-py3-none-any.whl
+```
+```text
+Archive:  dist/zipfpy-0.1-py3-none-any.whl
+  Length      Date    Time    Name
+---------  ---------- -----   ----
+      318  07-23-2019 20:24   zipfpy/__init__.py
+     1978  07-23-2019 20:24   zipfpy/check.py
+     2022  07-23-2019 20:24   zipfpy/generate.py
+      454  07-23-2019 20:30   zipfpy-0.1.data/scripts/check_zipf.py
+        7  07-23-2019 20:30   zipfpy-0.1.dist-info/top_level.txt
+       92  07-23-2019 20:30   zipfpy-0.1.dist-info/WHEEL
+      172  07-23-2019 20:30   zipfpy-0.1.dist-info/METADATA
+      606  07-23-2019 20:30   zipfpy-0.1.dist-info/RECORD
+---------                     -------
+     5649                     8 files
+```
+
+These distribution files can now be distributed via `pypi`, the standard Python
+repository for Python packages.
+
+There is a `testpypi` repository that lets us test putting packages
+on pypi without having them appear on the standard repository; let's
+start with that.  Go to [`https://test.pypi.org`](https://test.pypi.org),
+create an account, and confirm your email.
+
+The preferred tool for uploading packages to the PyPI family of repositories
+is called [`twine`](https://twine.readthedocs.io/en/latest/); we will
+install that package:
+
+```shell
+$ pip install twine
+```
+and now we can upload our distributions:
+
+```shell
+$ twine upload --repository-url https://test.pypi.org/legacy/ dist/*
+Enter your username: pterry
+Enter your passowrd: *********
+Uploading zipfpy-0.1-py3-none-any.whl
+...
+Uploading zipfpy-0.1.tar.gz
+...
+```
+
+The `--repository-url` allows us to specify the test PyPI repository, and we
+have now uploaded both types of distribution, allowing users with different
+environments to use the wheel distribution if their systems allow it, or
+the source distribution otherwise.
+
+We can now test this with a virtual environment and installing with pip:
+
+```shell
+$ virtualenv ~/envs/testpypi
+Using base prefix '/usr/local/Cellar/python/3.7.2_2/Frameworks/Python.framework/Versions/3.7'
+New python executable in /Users/pterry/envs/testpypi/bin/python3.7
+Also creating executable in /Users/pterry/envs/testpypi/bin/python
+sourInstalling setuptools, pip, wheel...
+$ source ~/envs/testpypi/bin/activate
+(testpypi) $ pip install --index-url https://test.pypi.org/simple/ zipfpy
+Looking in indexes: https://test.pypi.org/simple/
+Collecting zipfpy
+  Using cached https://test-files.pythonhosted.org/packages/f8/ac/d152eed95bff7567954becef22b8fe0f2aae15008fbeb346c69f75b9d2a0/zipfpy-0.1-py3-none-any.whl
+Installing collected packages: zipfpy
+Successfully installed zipfpy-0.1
+(testpypi) $
+```
+
+Success!  Once we are happy with how our package appears in testpypi
```

Comment:
Should we mention how to create conda packages? It is pretty straightforward once there already is a pypi package to use as a skeleton, but there are still caveats (such as different package names for some dependencies) and it is not as easy as with pip. I would lean towards skipping an explanation for now, but maybe including a link and mentioning it briefly.